### PR TITLE
Extend TitleIndex get_search_data to make a more specific optional se…

### PR DESCRIPTION
**Pull request by fmarco**
…arch inside specific pages placeholders plugins

This PR add a simple specialization of get_search_data, to handle generic cases or specific cases related to one or more pages.

The idea is to set up this variable inside the project settings

PLACEHOLDERS_SEARCH_LIST = {
                'page_title': [ 'placeholder_1', 'placeholder_2', etc. ],
            }

( or leave it empty to use the generic handling).

### CODE REVIEW

**Comment by czpython** [comment]
Hello @fmarco,

I really like the concept but I don't like coupling a value as dynamic as a title to a setting in the settings file because titles can easily change.

Also you could easily hit an issue with duplicate titles as well as having non-ascii in settings file if you would like to target titles in other pages.
 
And last, supporting multiple languages would be very repetitive, a placeholder is not bound to title instance which means that it won't change from one title to the next so coupling a title to placeholder slots would result in one of these per title per page.

I see two potential solutions:

* Do as django-cms does and make the setting global, meaning that we create a list of excluded placeholder slots.
* Instead of page_title we can use reverse_id which is less likely to change and will immediately apply to all titles under a page.

My personal opinion is that best approach from code/best practice point of view is point 1 but from a flexibility point of view then point 2.

We could also implement both solutions in a cool setting in order to provide an easy way to target all placeholders or a placeholder under page x.

**Comment by fmarco** [comment]
Hi @czpython , thanks for your answer!

I agree, page_title value is not so reliable by its nature, this first commit was a sort of proof of concept i thought during a project coding and i needed something similar.
I like the first point :), but maybe, as you said, we could leave the user free to use a safer settings as in point 1 or to use a more specific filter on revers_id.

**Comment by fmarco** [comment]
@czpython : Refactor of the code based on request and new tests structure

File: aldryn_search/search_indexes.py [hunk 1]
```diff
@@ -43,8 +43,49 @@ def get_plugin_queryset(self, language):
...
+        or leave it empty
+
+        PLACEHOLDERS_SEARCH_LIST = {}
+        """
         current_page = obj.page
-        placeholders = current_page.placeholders.all()
+        reverse_id = current_page.reverse_id
+        args = {}
+
+        try:
```
Comment (czpython) [inline aldryn_search/search_indexes.py]: maybe use getattr?

File: aldryn_search/tests.py [hunk 2]
```diff
@@ -96,9 +108,391 @@ def test_page_title_is_indexed_using_update_object(self):
...
+        self.request = get_request(language='en')
+
+    def test_page_title_is_indexed_using_prepare_with_filter_option(self):
+        """This tests the indexing path way used by update_index mgmt command"""
+        page = create_page(title="test_page", reverse_id='testpage', template="test.html", language="en")
+        plugin = add_plugin(page.placeholders.get(slot='content'), NotIndexedPlugin, 'en')
+        plugin2 = add_plugin(page.placeholders.get(slot='hidden_content'), HiddenPlugin, 'en')
+
+
+        from haystack import connections
```
Comment (czpython) [inline aldryn_search/tests.py]: best to create a base test case and provide something like get_title_index() which can have the following:

```python
def get_title_index(self):
    search_conn = connections[DEFAULT_ALIAS]
    unified_index = search_conn.get_unified_index()
    index = unified_index.get_index(Title)
    return index
```

File: aldryn_search/search_indexes.py [hunk 3]
```diff
@@ -43,8 +43,49 @@ def get_plugin_queryset(self, language):
...
+        'slot_name' -> search for slot_name
+        '-slot_name' -> exclude slot name from the search list
+
+        or leave it empty
+
+        PLACEHOLDERS_SEARCH_LIST = {}
+        """
         current_page = obj.page
-        placeholders = current_page.placeholders.all()
+        reverse_id = current_page.reverse_id
```
Comment (czpython) [inline aldryn_search/search_indexes.py]: please move the logic from here all the way to the placeholder queryset to a method called get_page_placeholders, which takes a page as parameter and returns placeholder queryset after doing the filtering logic

File: aldryn_search/search_indexes.py [hunk 4]
```diff
@@ -43,8 +43,49 @@ def get_plugin_queryset(self, language):
...
+
+        if placeholders_by_page:
+            filter_target = None
+            if '*' in placeholders_by_page:
+                filter_target = '*'
+            if reverse_id and reverse_id in placeholders_by_page:
+                filter_target = reverse_id
+            if not filter_target:
+                raise AttributeError('Leave PLACEHOLDERS_SEARCH_LIST empty or set up at least the generic handling')
+            slots = [el for el in placeholders_by_page[filter_target] if not el.startswith('-')]
```
Comment (czpython) [inline aldryn_search/search_indexes.py]: considering ```-something``` can easily be a valid placeholder slot name, I suggest we use a different character either ```!slotname``` or maybe just use a dict with ```{'include': [], 'exclude': []}```, the dictionary approach to me seems more explicit.

Comment (fmarco) [inline aldryn_search/search_indexes.py]: I agree, i'm thinking something similar, i like a lot this solution!

**Comment by fmarco** [comment]
Fixed

**Comment by czpython** [comment]
@fmarco Thank you. I will merge this now, but please make another pr for https://github.com/nephila/aldryn-search/blob/feature/improve_titleindex_get_search_data/aldryn_search/tests.py#L306 seems like this test is repeated

**Pull request merged**